### ISSUE TITLE
[semantic highlight] Add basic language tokens

### ISSUE THREAD

**Issue by sclu1034** [clangd/clangd#1115]
Looking at `HighlightingKind`, clangd's semantic highlighting seems to be missing a lot of basic tokens (e.g. strings, numbers, operators). Because of that, highlighting based on this alone looks very barren:

![image](https://user-images.githubusercontent.com/4508454/164003167-1f7291eb-d635-45a7-af16-e2fd15f8f8f4.png)

VSCode doesn't look as broken, because it seems to combine semantic highlighting with its regular textmate highlighters. But it would be nice if highlighting with clangd only would be possible.

For reference, Rust Analyzer has a [fairly extensive list of tokens/modifiers](https://github.com/rust-lang/rust-analyzer/blob/master/editors/code/package.json#L1143-L1422).

**Comment by HighCommander4** [comment]
Based on previous discussions, e.g. https://github.com/clangd/clangd/issues/352#issuecomment-619442147:

> Things that are easily lexically derived (such as preprocessor defines and doxygen tags) don't need to be in the language server I think. At least, we haven't tried so far to make the coloring exhaustive, because we assume it's always a layer over syntax highlighting. Compared to syntax highlighting, LSP based syntax highlighting is asynchronous and imposes extra costs like serialization.

it seems unlikely that we're going to do this in clangd.

**Comment by sclu1034** [comment]
> we assume it's always a layer over syntax highlighting

I see.

The flip-side is that in order to get this "lower layer syntax highlighting" an editor can either
- pull in a second instance of the dedicated language parser, e.g. tree sitter, which then has the same latency 'issues' as clangd, but cannot share with it
- have a generic, approximate parsing system (textmate, regex, ...) which, in my experience, doesn't get to the same level as a dedicated parser

But I guess I'll have to look into implementing tree-sitter syntax highlighting for this, then.

**Comment by HighCommander4** [comment]
>  * have a generic, approximate parsing system (textmate, regex, ...) which, in my experience, doesn't get to the same level as a dedicated parser

In my experience, the C++ TextMate grammar at https://github.com/jeff-hykin/better-cpp-syntax (used by vscode) is pretty accurate for the things clangd's semantic tokens don't cover.

There are a few template edge cases like https://github.com/jeff-hykin/better-cpp-syntax/issues/370 that it gets wrong, but they're actually caused by that grammar trying **too** hard (wanting to provide reasonable heuristic colors for identifiers for non-LSP use cases).

I'm pretty sure a C++ TextMate grammar designed to only highlight non-semantic token types could nail it and be very accurate.

> * pull in a second instance of the dedicated language parser, e.g. tree sitter, which then has the same latency 'issues' as clangd, but cannot share with it

IMO tree-sitter is overkill in comparison to TextMate, but tree-sitter will still have better latency than clangd's semantic tokens, at least on file-open, since by design it does not need to process the contents of #included files.

**Comment by HighCommander4** [comment]
> Do you have tests/estimates of how much additional tokens "cost"? Rust Analyzer provides tokens for pretty much every character in a file, and it seems to be doing just fine on performance in my experience.

I don't; it would be interesting to see some numbers.

I personally have no objection to giving clangd an option to provide semantic tokens for all C++ syntax. From an implementation point of view, I expect the missing tokens are all "lexical" (i.e. correspond to tokens in `TokenBuffer`, with token types no more granular than C++ lexical token types), and so could be provided relatively easily. May want to run this by @sam-mccall before spending time on it though.

**Comment by daiyousei-qz** [comment]
> > we assume it's always a layer over syntax highlighting
> 
> I see.
> 
> The flip-side is that in order to get this "lower layer syntax highlighting" an editor can either
> 
> * pull in a second instance of the dedicated language parser, e.g. tree sitter, which then has the same latency 'issues' as clangd, but cannot share with it
> * have a generic, approximate parsing system (textmate, regex, ...) which, in my experience, doesn't get to the same level as a dedicated parser
> 
> But I guess I'll have to look into implementing tree-sitter syntax highlighting for this, then.

I have an experimental vscode extension [Minimal C++ Syntax](https://marketplace.visualstudio.com/items?itemName=daiyousei-qz.minimal-cpp-syntax) which does exactly what @HighCommander4 mentioned above, which works pretty well for me (though it's not that mature and well-tested). Excluding semantic tokens that are already colored by the clangd, the remaining tokens are pretty easy to recognize with regex. Basically, what we need to match are mostly context-insensitive tokens like comments, strings, constants, keywords and .etc, which could be done with a few hundred lines of textmate rules.

I find that the pain points are preprocessors, especially function-like macros, and obfuscated program with backslash or other preprocessing tricks. But they are really minor in day-to-day sources.

**Comment by sclu1034** [comment]
> I have an experimental vscode extension [Minimal C++ Syntax](https://marketplace.visualstudio.com/items?itemName=daiyousei-qz.minimal-cpp-syntax)

That's great for you, but doesn't help me or lots of other people. Going through this thread, I always see "VScode has this", "VSCode can do that". But VSCode is not the only text editor with LSP support out there. And not all of them support TextMate.

As far as I'm aware, the reason why there is a "P" in LSP is because its goal is to provide IDE-like features to _any_ editor that has a client for it, while also making features and configuration uniform across editors. And syntax highlighting is one of those features that the protocol supports.

But the way clangd currently implements it requires every one of those editors to _also_ implement syntax highlighting. And every theme creator/user has to do extra work to configure both and make them play nicely together.

**Comment by HighCommander4** [comment]
I agree that clangd should follow the LSP, and cater to all clients not just VSCode (and we do try -- we have users happily using other clients such as vim, emacs, Sublime Text, and others).

However, the name of the coloring feature in the LSP is "**semantic** tokens", and its [description](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_semanticTokens) is as follows (emphasis mine):

> Semantic tokens are used to add **additional** color information to a file **that depends on language specific symbol information**.

My understanding is that this has been fairly widely interpreted in the LSP community to mean color information that requires the full semantic model built by a language server (e.g. classification of identifier tokens as types vs. variables), which are applied on top of client-side coloring of basic language syntax such as keywords.

It's possible I'm off base here, and perhaps it's worth seeking clarification in the [LSP repo](https://github.com/microsoft/language-server-protocol). If the LSP maintainers clarify that the intention is for `semanticTokens` to cover all tokens in a file, I think that would be a strong argument for clangd to do so by default.

Otherwise, my personal opinion remains that it may be appropriate for clangd to have an option to do this, but should not do it by default.

**Comment by sclu1034** [comment]
Thank you for the quote, I wasn't aware of that specific wording.

But in the same section, the "predefined values" already include some of the token types I felt missing in clangd, such as `string` or `number`, which would be some of those "just syntax" kind.

As for the LSP maintainers, I just found an existing discussion at https://github.com/microsoft/language-server-protocol/issues/1063, but without response.

An off-by-default option in clangd would be fine for my needs.

**Comment by sclu1034** [comment]
I just noticed: The upcoming spec version has a capability that's very close to the mentioned option:

```ts
/**
 * Whether the client uses semantic tokens to augment existing
 * syntax tokens. If set to `true` client side created syntax
 * tokens and semantic tokens are both used for colorization. If
 * set to `false` the client only uses the returned semantic tokens
 * for colorization.
 *
 * If the value is `undefined` then the client behavior is not
 * specified.
 *
 * @since 3.17.0
 */
augmentsSyntaxTokens?: boolean;
```

**Comment by sam-mccall** [comment]
I'd be happy to see this in as an option if someone wants to contribute it. On the other hand I/others can't really prioritize it if no widely used editors will benefit. So I think we call this one "patches welcome".

I agree `augmentsSyntaxTokens` sounds exactly like the option we need, so I don't think we need to add an explicit flag. 
When `augmentsSyntaxTokens` is unset we should keep the current behavior.

Potential costs are:
 - server-side complexity: likely to be smallish as we can basically just dispatch on token-type, right?
 - server-side performance: I doubt 3x more tokens or so will make a difference
 - client-side performance: manipulating large highlight sets for big files can be really slow (I know this is true in vim). big JSON-wrangling can be somewhat slow too in (bad) extension languages.
 - bandwidth: we and others run browser-based monaco with clangd, so request/response sizes matter.
 - maintenance: there's bound to be edge-cases or disagreements about how we classify tokens.

Only the last is really a problem if editors are opting in.

**Comment by HighCommander4** [comment]
> I just noticed: The upcoming spec version has a new client capability that's very close to the mentioned option:

Nice find, I think this makes the spec's intention clear.

I gave a suggestion for a high-level implementation approach in a [previous comment](https://github.com/clangd/clangd/issues/1115#issuecomment-1104350159):

> From an implementation point of view, I expect the missing tokens are all "lexical" (i.e. correspond to tokens in TokenBuffer, with token types no more granular than C++ lexical token types), and so could be provided relatively easily.

Happy to provide more detailed guidance if someone is interested in working on this.

**Comment by daiyousei-qz** [comment]
@HighCommander4 Is this being worked on by anybody? If not, I can spend some time on this.

Here's my random ideas:
1. As the missing tokens are lexical and well-defined by itself or its immediate surroundings, we should have a dedicated scanner which directly operates source text instead of AST. The semantic tokens could be computed from the scanner and merged into current semantic highlighter. Maybe this could run in parallel with the semantic highlighter.
2. To reduce the amount of redundant computation, we should implement `provideDocumentRangeSemanticTokens` so those lexical tokens could be computed within a small range of source code instead of the entire source. My understanding on the current semantic highlight implementation is that we iterate the entire AST and trim the tokens by doing diff. Am I correct?
3. We should still leave some tokens to syntax highlight to allow users to take advantage of language injection. For example, people may write glsl/hlsl shader code in string literals. And some extension may color text in comments accordingly. An option is to have each token type configurable.

**Comment by HighCommander4** [comment]
> @HighCommander4 Is this being worked on by anybody?

Not that I'm aware of.

>     1. As the missing tokens are lexical and well-defined by itself or its immediate surroundings, we should have a dedicated scanner which directly operates source text instead of AST. The semantic tokens could be computed from the scanner and merged into current semantic highlighter. Maybe this could run in parallel with the semantic highlighter.

Agree that we don't need the AST to provide the missing tokens. We have the token stream from the lexer available as a [TokenBuffer](https://searchfox.org/llvm/rev/9ce232fba99c47c3246f06fcbe37c24b9d90585f/clang-tools-extra/clangd/SemanticHighlighting.cpp#470), I would recommend using that. I believe providing lexical highlighting tokens should be a fairly simple matter of iterating through the tokens in the TokenBuffer, selecting the subset for which we want to provide a lexical highlighting, and performing a simple mapping from the clang token kind to the highlighting kind.

>     2. To reduce the amount of redundant computation, we should implement `provideDocumentRangeSemanticTokens` so those lexical tokens could be computed within a small range of source code instead of the entire source. My understanding on the current semantic highlight implementation is that we iterate the entire AST and trim the tokens by doing diff. Am I correct?

LSP provides two different ways to reduce the overhead of semantic tokens requests:

 1. "Delta" requests, where the server only sends the diff between the last request and the current one, and the client applies this diff to the result of the last request which it retains.
 2. "Range" requests, where the client only requests tokens for a subset of the file (presumably, the portion currently visible) and the server only sends tokens in this range.

Both are optional and will be used if available. Clangd supports (1) but not (2) (I believe we felt (2) would not bring a significant performance improvement for the complexity.) I don't think this needs to change for providing lexical tokens.

>     3. We should still leave some tokens to syntax highlight to allow users to take advantage of language injection. For example, people may write glsl/hlsl shader code in string literals. And some extension may color text in comments accordingly. An option is to have each token type configurable.

I don't have a strong opinion on this; having knobs to disable particular tokens like string literals to allow another plugin to provide finer-grained tokens for their contents (like your mentioned shader example) sounds fine to me.

**Comment by HighCommander4** [comment]
> I believe providing lexical highlighting tokens should be a fairly simple matter of iterating through the tokens in the TokenBuffer, selecting the subset for which we want to provide a lexical highlighting, and performing a simple mapping from the clang token kind to the highlighting kind.

I think LSP expects tokens to be on a single line, so we'll probably need to split multi-line tokens into multiple single-line tokens.

**Comment by sclu1034** [comment]
> I think LSP expects tokens to be on a single line, so we'll probably need to split multi-line tokens into multiple single-line tokens.

It's up to the client to support (and report) that:

> Whether a token can span multiple lines is defined by the client capability `multilineTokenSupport`. If multiline tokens are not supported and a tokens length takes it past the end of the line, it should be treated as if the token ends at the end of the line and will not wrap onto the next line.
>
> https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_semanticTokens

**Comment by daiyousei-qz** [comment]
> > 1. As the missing tokens are lexical and well-defined by itself or its immediate surroundings, we should have a dedicated scanner which directly operates source text instead of AST. The semantic tokens could be computed from the scanner and merged into current semantic highlighter. Maybe this could run in parallel with the semantic highlighter. 
> Agree that we don't need the AST to provide the missing tokens. We have the token stream from the lexer available as a [TokenBuffer](https://searchfox.org/llvm/rev/9ce232fba99c47c3246f06fcbe37c24b9d90585f/clang-tools-extra/clangd/SemanticHighlighting.cpp#470), I would recommend using that. I believe providing lexical highlighting tokens should be a fairly simple matter of iterating through the tokens in the TokenBuffer, selecting the subset for which we want to provide a lexical highlighting, and performing a simple mapping from the clang token kind to the highlighting kind.
>

Yes, it would be great if spelled TokenBuffer could be reused for this purpose. However, there's a problem with the approach because of preprocessor. It seems raw lexer will provide incorrect token type for preprocessing directives and eliminate the boundary of that, which makes it hard to highlight the preprocessors.

> > 2. To reduce the amount of redundant computation, we should implement `provideDocumentRangeSemanticTokens` so those lexical tokens could be computed within a small range of source code instead of the entire source. My understanding on the current semantic highlight implementation is that we iterate the entire AST and trim the tokens by doing diff. Am I correct?
> 
> LSP provides two different ways to reduce the overhead of semantic tokens requests:
> 
> 1. "Delta" requests, where the server only sends the diff between the last request and the current one, and the client applies this diff to the result of the last request which it retains.
> 2. "Range" requests, where the client only requests tokens for a subset of the file (presumably, the portion currently visible) and the server only sends tokens in this range.
> 
> Both are optional and will be used if available. Clangd supports (1) but not (2) (I believe we felt (2) would not bring a significant performance improvement for the complexity.) I don't think this needs to change for providing lexical tokens.

I experimented with a hack into clangd to support range request. it looks it's not worth it to support the range request because:
1. The major bottleneck of a semantic token request is waiting for building AST instead of traversal.
2. The client may not take advantage of range request at all. VSCode will only issue a couple of range requests at opening the file. After that, only full request will be sent as long as it's supported.
 
---

Here, I'd like to list and clarify the features that may/may not be wanted.

### Lexical Tokens
There are roughly the following lexical tokens in the language
- comments
- keywords
- number literal
- char literal
- string literal
- label

Furthermore, it might be helpful to add modifier to refine a lexical token for better theming:
![image](https://user-images.githubusercontent.com/12378019/173171582-b6627591-5f2f-42c5-bd32-fa4f801148be.png)

### Preprocessor
Preprocessor definition is technically not that lexical because its definition is more structured. However, since they are part of the language (or a meta language on its own), I expect they should be properly highlighted as well. Including:
- preprocessor directive
- pragma commands
- quoted/unquoted string
- defined(...), __has_include(...) and .etc
- line-continuation symbol
- ...

And maybe we could implement something regex-based syntax highlight cannot do. Like highlight macro parameters in the macro definition (can X in "X::Y()" be assigned with a semantic token?).
![image](https://user-images.githubusercontent.com/12378019/173172446-b3da18d3-a207-48f7-ba4f-73a20fb5f27e.png)

Notably, those are features that can't be easily implemented with the existing TokenBuffer.

**Comment by HighCommander4** [comment]
> ### Lexical Tokens
> 
> There are roughly the following lexical tokens in the language
> 
>     * comments
> 
>     * keywords
> 
>     * number literal
> 
>     * char literal
> 
>     * string literal
> 
>     * label
> 
> 
> Furthermore, it might be helpful to add modifier to refine a lexical token for better theming: ![image](https://user-images.githubusercontent.com/12378019/173171582-b6627591-5f2f-42c5-bd32-fa4f801148be.png)

This sounds reasonable to me, and should be implementable using `TokenBuffer`.

> ### Preprocessor
> 
> Preprocessor definition is technically not that lexical because its definition is more structured. However, since they are part of the language (or a meta language on its own), I expect they should be properly highlighted as well. Including:
> 
>     * preprocessor directive
> 
>     * pragma commands
> 
>     * quoted/unquoted string
> 
>     * defined(...), __has_include(...) and .etc
> 
>     * line-continuation symbol
> 
>     * ...

For this, I think we can use clangd's existing [`PPCallbacks`](https://searchfox.org/llvm/rev/8daaea206b5905858bdbdd881be4fe9bdcc54516/clang-tools-extra/clangd/CollectMacros.h#44). We can store any additional information we need for highlighting (e.g. the bodies of macro definitions) in `MainFileMacros`, which is then available in the semantic highlighter as [`AST.getMacros()`](https://searchfox.org/llvm/rev/8daaea206b5905858bdbdd881be4fe9bdcc54516/clang-tools-extra/clangd/SemanticHighlighting.cpp#846).

> And maybe we could implement something regex-based syntax highlight cannot do. Like highlight macro parameters in the macro definition (can X in "X::Y()" be assigned with a semantic token?). ![image](https://user-images.githubusercontent.com/12378019/173172446-b3da18d3-a207-48f7-ba4f-73a20fb5f27e.png)

This sounds useful, and it should be possible. High-level sketch:

 * Given a `MacroInfo` representing a macro definitions, loop through its [`ReplacemenTokens`](https://searchfox.org/llvm/rev/8daaea206b5905858bdbdd881be4fe9bdcc54516/clang/include/clang/Lex/MacroInfo.h#58)
 * If it's an identifier token, check if it corresponds to a macor parameter by calling [`getParameterNum(IdentifierInfo*)`](https://searchfox.org/llvm/rev/8daaea206b5905858bdbdd881be4fe9bdcc54516/clang/include/clang/Lex/MacroInfo.h#192) and if so highlight it

**Comment by sclu1034** [comment]
> ### Lexical Tokens
> 
> There are roughly the following lexical tokens in the language
> 
>     * comments
> 
>     * keywords
> 
>     * number literal
> 
>     * char literal
> 
>     * string literal
> 
>     * label
> 

Operators and punctuation should be added as well.
